go-tutorial.md

```diff
@@ -0,0 +1,1233 @@
+- Debugging
+
+The bazel open-source community also provides another tool called [gazelle](https://github.com/bazelbuild/bazel-gazelle).
+Gazelle addresses the creation and maintence of [BUILD](https://bazel.build/concepts/build-files) files.  
+Every bazel project has BUILD (BUILD.bazel) files that define the various rules that are used within a project.
+When you add more code or dependencies to a project you need to update your build files.  When you add a new folder
+you need to add another BUILD file. If you have ever worked with bazel you know how much time you spend maintaining
+these files, if you maintain the files by hand.  Gazelle was created to reduce the previously mentioned pain points.
+
+> Gazelle is a build file generator for Bazel projects. It can create new BUILD.bazel files for a project that follows language conventions, and it can update existing build files to include new sources, dependencies, and options. Gazelle natively supports Go and protobuf, and it may be extended to support new languages and custom rule sets.
+>
+>  -- <cite>https://github.com/bazelbuild/bazel-gazelle#gazelle-build-file-generator</cite> 
+
+Intially gazelle was created to support Go, and now supports many other languages.
+
+A lot of understand bazel well is understanding the configuration language that bazel uses.
+The language is called [StarLark](https://github.com/bazelbuild/starlark).
+
+> Starlark (formerly known as Skylark) is a language intended for use as a configuration language. It was designed for the Bazel build system, but may be useful for other projects as well. This repository is where Starlark features are proposed, discussed, and specified. It contains information about the language, including the specification. There are multiple implementations of Starlark.
+>
+> Starlark is a dialect of Python. Like Python, it is a dynamically typed language with high-level data types, first-class functions with lexical scope, and garbage collection.
+> - <cite>https://github.com/bazelbuild/starlark#overview</cite>
+
+The good news is that Starlark is a dialect of Python, almost a subset of the language.  If you know
+Python you have a jump start on learning Starlark.
+
+Before we start going through rules_go and gazelle we are going to cover a couple of dependencies for this 
+tutorial and create a simple Go project.
+
+## Dependencies for the tutorial
+
+We use the following dependencies for this tutorial.
+
+- go: https://go.dev/doc/install
+- gcc: use your systems package manager
+- bazelisk: https://github.com/bazelbuild/bazelisk#installation
+
+Technically we do not need the go binary installed, but we are going to use
+`cobra-cli` to generate some project code.  We did not want to add the 
+extract work in bazel to run the binary using bazel. A developer, using go,
+does not need to download the go binary. In order to keep a build deterministic
+bazel and rules_go download go. rules_go require that gcc is installed.
```

Comment:
I get not wanting to set up `cobra-cli` as a dep of the Bazel workspace. What do you think about using the Go SDK downloaded by `rules_go` to keep things reasonably hermetic?

The steps would be something like:
1. Create the `WORKSPACE` as you have described, except do not include the loading of the `deps.bzl` file.
2. Init the module: `bazel run @go_sdk//:bin/go -- mod init github.com/chrislovecnm/bazel-go-example-code`
3. Install `cobra-cli`: `bazel run @go_sdk//:bin/go -- install github.com/spf13/cobra-cli@latest`
4. Perform your `cobra-cli` steps: `$HOME/go/bin/cobra-cli init -a '${NAME}' --license apache`

All of that being said, I would be amazed that someone walking through the tutorial does not already have `go` installed. What I think is interesting about using the downloaded SDK is that it shows off the hermetic toolchain.

Just throwing it out there.

Reply:
I think it is pretty cool, yes it shows how to use the go rules more.  Let me see how it works out.  I would like to make this an issue and update the tutorial after we get this merged.  Thoughts?

Reply:
That is fine with me.